### PR TITLE
Support C++14

### PR DESCRIPTION
**Pull request by drwells**
I tried to write a generic lambda function yesterday and ended up rewriting a large part of the compiler feature detection code (quite an adventure in yak shaving). I made a lot of changes. Some highlights:

~~1. I renamed `DEAL_II_WITH_CXX11` to `DEAL_II_WITH_AT_LEAST_CXX11` in some contexts so that, when using language standards that are supersets of C++11 (such as C++14, modulo some corner cases that do not apply to current deal.II code), the library is preprocessed correctly. The old macro name would certainly work, but I think the seperate name is clearer. This macro is defined when compiling in either C++14 or C++11 mode. I still define the old macro whenever the new one is defined (see `include/deal.II/base/config.h.in`) for backwards compatability in application code.~~ Edit: this choice is overly pedantic and I undid it.
2. I renamed `DEAL_II_CXX11_FLAG` to `DEAL_II_CXX_VERSION_FLAG` for similar reasons. The old flag is still accepted (see `cmake/checks/check_01_cxx_features.cmake`).
3. CMake now looks for the boolean `DEAL_II_WITH_CXX14` during configuration, in addition to `DEAL_II_WITH_CXX11`. Edit: Both are exported (see `include/deal.II/base/config.h.in`).
4. I updated some of the documentation to more accurately describe the C++11 and C++14 compilation options.

I tried to make the changes general enough so that, some day, deal.II may also support C++17 without much more work.

I am far from an expert in CMake, so I would appreciate any feedback or criticism.Things seem to work with CMake `3.2.3` on my machine.

Finally, it is worth noting that one could obtain the same effect with the unintuitive naming trick `-DDEAL_II_CXX11_FLAG="-std=c++14"`.

Edit: I force updated with @tamiko's suggestions.

### CODE REVIEW

**Comment by tamiko** [comment]
I'll have a closer look at the CMake stuff next week.

Just one remark. I do not like the naming <code>DEAL_II_WITH_AT_LEAST_CXX11</code>.
What about simply having <code>DEAL_II_WITH_CXX11</code> and <code>DEAL_II_WITH_CXX14</code>?

We already have feature dependencies, so the implication "if <code>DEAL_II_WITH_CXX14</code> is selected, <code>DEAL_II_WITH_CXX11</code> has to be enabled, too" nicely fits into this category. Further, just keeping the old name frees us from supporting legacy feature flags.

The idea to select an appropriate <code>DEAL_II_CXX_VERSION_FLAG</code> is a nice generalization.

File: cmake/checks/check_01_cxx_features.cmake [hunk 1]
```diff
@@ -254,19 +329,39 @@ IF(DEAL_II_WITH_CXX11 AND NOT DEAL_II_HAVE_CXX11)
...
+    "C++14 support was requested (DEAL_II_WITH_CXX14=${DEAL_II_WITH_CXX14}) "
+    "but is not supported by the current compiler and flag combination.\n"
+    "Please disable C++14 support, i.e. configure with\n"
+    "    -DDEAL_II_WITH_CXX14=FALSE,\n"
+    "or use a different compiler, instead. (If the compiler flag for C++14 "
+    "support differs from \"-std=c++14\", a suitable "
+    "compiler flag has to be specified manually via\n"
+    "    -DDEAL_II_CXX_VERSION_FLAG=\"...\"\n\n"
     )
 ENDIF()
```
Comment (tamiko) [inline cmake/checks/check_01_cxx_features.cmake]: Perfect. This is exactly what I had in mind.

File: cmake/checks/check_01_cxx_features.cmake [hunk 2]
```diff
@@ -28,29 +29,71 @@
 
 ########################################################################
 #                                                                      #
-#                             C++11 Support:                           #
+#                         C++ Version Support:                         #
 #                                                                      #
 ########################################################################
 
+#
+# backwards compatability with the old DEAL_II_CXX11_FLAG option
```
Comment (tamiko) [inline cmake/checks/check_01_cxx_features.cmake]: * compatibility :-)

Comment (drwells) [inline cmake/checks/check_01_cxx_features.cmake]: The worst part of that is I just fixed "compatbile" to "compatible" in another file. English is too hard :(

**Pull request merged**